python/cudf/cudf/utils/ioutils.py

```diff
@@ -780,12 +780,14 @@ def get_filepath_or_buffer(path_or_data, compression, iotypes=(BytesIO)):
     """
     if is_url(path_or_data):
         with urllib.request.urlopen(path_or_data) as url:
-            compression = url.headers.get('Content-Encoding', None)
+            compression = url.headers.get("Content-Encoding", None)
             buffer = BytesIO(url.read())
         return buffer, compression
     elif isinstance(path_or_data, str):
         return os.path.expanduser(path_or_data), compression
     elif not isinstance(path_or_data, iotypes) and is_file_like(path_or_data):
+        if isinstance(path_or_data, TextIOWrapper):
+            path_or_data = path_or_data.buffer
```

Comment:
@mrocklin @quasiben @j-ieong Seems like Dask is now returning an `io.TextIOWrapper` object for filepaths in certain situations which the `read()` function returns a string instead of bytes. According to the docs using this `buffer` attribute isn't best practice, but encoding a string to bytes is super expensive. Should Dask not use an `io.TextIOWrapper` and instead use something else that returns bytes here?

Reply:
Yeah, can Dask avoid using `io.TextIOWrapper` if possible? I would think it's only needed if the target can't handle bytes and can only handle strings...

The readers would prefer something that is like a Python `bytes` object or `bytesarray`, since the Cython layer can simply pass it straight to the C++ layer.

Reply:
In which situations does this occur?  This is for read_csv and such?  How did you get to this point?

The relevant code in Dask seems to be here: 

https://github.com/dask/dask/blob/140a27b57dfe0c3a5a694e04c3c0cd07412d56eb/dask/bytes/core.py#L221-L240

We seem to only use the TextIOWrapper if the user asks for text data.  

If this is in read_csv I think that this uses the `dask.bytes.core.read_bytes` function, which seems to default to `"rb"` mode.  

https://github.com/dask/dask/blob/140a27b57dfe0c3a5a694e04c3c0cd07412d56eb/dask/bytes/core.py#L189-L198

So, at least on first pass I don't see why you would be getting a TextIOWrapper.  If you can provide more context that would be helpful.

Reply:
@mrocklin I was getting it on all of the `read_json` dask-cudf tests from the looks of it. Those tests write a json file using Pandas or Dask and then try to read it using the cudf json reader as the engine underneath Dask. See `dask_cudf/io/tests/test_json.py`

Reply:
Yup, looks like that's here: 

https://github.com/dask/dask/blob/140a27b57dfe0c3a5a694e04c3c0cd07412d56eb/dask/dataframe/io/json.py#L203-L210

@martindurant, do you know why we need to decode the bytes in read_json?

Reply:
It's been a while... I assume that the reader engine expects text as opposed to bytes, but maybe that was only true in py2.

Reply:
I'm going to merge this as is, but if you want me to open an issue against Dask just let me know 😄